File: python/LC_3_LongestSubstringWithoutRepeatChars.py
# https://leetcode.com/problems/longest-substring-without-repeating-characters
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        chars_set = set()
        left, right, maxLen, best_start = 0, 0, 0, 0
        while right < len(s):
            if s[right] not in chars_set:
                chars_set.add(s[right])
                if right - left + 1 > maxLen:
                    maxLen = right - left + 1
                    best_start = left
                right += 1
            else:
                chars_set.remove(s[left])
                left += 1
        print(s[best_start : best_start+maxLen])
        return maxLen

sol = Solution()
assert sol.lengthOfLongestSubstring("abcabcbb") == 3
assert sol.lengthOfLongestSubstring("bbbbb") == 1
assert sol.lengthOfLongestSubstring("pwwkew") == 3

File: java/src/sliding_window/LC_3_LongestSubstringWithoutRepeatChars.java
package sliding_window;
//  https://leetcode.com/problems/longest-substring-without-repeating-characters

import java.util.HashSet;

public class LC_3_LongestSubstringWithoutRepeatChars {

       public void lengthOfLongestSubstring(String s) {
        int left = 0, right = 0, maxLength = 0;
        HashSet<Character> charactersSet = new HashSet<>();
        while (right < s.length()) {
            char currentChar = s.charAt(right);
            if (!charactersSet.contains(currentChar)) {
                charactersSet.add(currentChar);
                maxLength = Math.max(maxLength, right - left + 1);
                right++;
            } else {
                charactersSet.remove(s.charAt(left));
                left++;
            }
        }
        System.out.println(maxLength);
    }

    public static void main(String[] args) {
        LC_3_LongestSubstringWithoutRepeatChars solution = new LC_3_LongestSubstringWithoutRepeatChars();
        solution.lengthOfLongestSubstring("abcabcbb");
        solution.lengthOfLongestSubstring("bbbbbb");
        solution.lengthOfLongestSubstring("pwwkew");
    }
}
